### PR TITLE
[core][pruning][feature] cuSPARSELt 2:4 sparse bindings

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97546
* #102133

This PR integrates cuSPARSELt v0.4.0.7 into pytorch.

It is composed of two elements:
1. A torch custom class that is used to store and manage the cusparselt
constructs needed to do sparse matrix multiplication
2. A tensor subclass that overrides the dispatch of torch.t(),
torch.mm() and torch.addmm() to use the cusparselt sparse matmul and
also store the custom class state.

For performance and memory overhead reasons, we'd like to cache the descriptors and
compressed matrix that are used in cusparselt. However this makes it a
bit tricky, since this means there's some state that we have to manage.

Previously, we were holding this state in a cuSPARSELtLinear module and
swapping that module with nn.Linear. This works fine for Linear, since
the forward() function is just an addmm, but doesn't work great when
expanding to modules that have a more complicated forward() function,
since we need to copy over all the custom logic.

With tensor subclasses, we can store the state on the tensor itself, and
then at dispatch time retrieve it from the tensor. This essentially
defines a custom matmul function for each tensor.

Additionally, conceptually cusparselt matmul is closer to
torch.addmm/torch.mm so it makes more sense to do the replacement at
that level. It also leads to a cleaner UX, where previously a user had
to use our pruning flow e2e in order to utilize `convert`, now all they
have to do is get their weights into a 2:4 dense format (with 0s) and
then all they have to do to get accelerated inference is

```
from torch.ao.pruning import SemiStructuredSparseTensor
model = Model()

model.linear.weight = nn.Parameter(SemiStructuredSparseTensor(model.linear.weight))

```

Our pruning flow has functionality to get the weights in this format by
using `pruner.squash_mask()`

I've also added an addtional `contiguous_output` flag which lets us fuse
a subsequent transpose operation into the cusparselt matmul call. This
is especially useful for distributed settings, since the output of our
cusparselt matmul is Transposed and that messes up the collect/gather.

With contiguous_output set to True, the output will be contiguous,
meaning it should perfectly match F.linear and can be used as a drop-in
replacement in distributed settings. You can see an example of how to
use it below.

```
from torch.ao.pruning import SemiStructuredSparseTensor
model = Model()

model.linear.weight = nn.Parameter(SemiStructuredSparseTensor(model.linear.weight, contiguous_output=True))

```

However there are some kinks within the workflow we still
need to resolve.

1. Integration with CUTLASS - sometimes CUTLASS can be faster and
supports more advanced epilogue fusions, like SwigLu, so we need some
way of deciding when to dispatch into CUTLASS vs CUSPARSELT.

2. No way to go from compressed matrix to sparse matrix - Currently
this means we do not support .t() fully on SemiStructuredSparseTensors,
since we can't transpose the compressed form. We need some functionality
to go back to the dense representation from the compressed form. CUTLASS
has this ability and we may be able to reuse their implementation if
they share the same meta/mask layout.

3. Bias propogating wrong way. For when the sparse matrix is first in
addmm, the bias is propogated columnwise instead of rowwise. Note that this is
fine when the second matrix is dense, since we transpose the result
anyways.

4. Padding - currently cusparselt only supports dimensions that are
multiples of 16, 8, 4 (depending on dtype), so we should add padding so
that we can satisfy this constraint for all matrices.

dtypes supported:
- int8
- fp16
- bf16
- fp32

Ops supported:
```
torch.addmm(bias, dense, sparse)
torch.addmm(bias, sparse, dense)
torch.addmm(dense, sparse)
torch.addmm(sparse, dense)
```

### CODE REVIEW

**Comment by jcaip** [comment]
@cpuhrsch Would you mind reviewing the PR? You can ignore the .py files, I'm mainly just looking for any issues with the CUDA kernels / C++ code.

File: .env.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# faster build
```
Comment (cpuhrsch) [inline .env.sh:1]: Seems like an accidentally added file?

File: aten/src/ATen/CMakeLists.txt [hunk 2]
```diff
@@ -439,9 +439,10 @@ if(USE_CUDA AND NOT USE_ROCM)
   if($ENV{ATEN_STATIC_CUDA})
     list(APPEND ATen_CUDA_DEPENDENCY_LIBS
       ${CUDA_LIBRARIES}
-      CUDA::cusparse_static
```
Comment (cpuhrsch) [inline aten/src/ATen/CMakeLists.txt:442]: Why do you need to touch the other files too and why does the format need to change?

File: aten/src/ATen/native/cuda/CusparseLtKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
+/*
+The following source file implements a sparse linear operator using cusparseLt
+*/
+#include <torch/custom_class.h>
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/CusparseLtKernels.cpp:4]: We can't have dependencies on `torch` within aten.

File: aten/src/ATen/native/cuda/CusparseLtKernels.cpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+#include <torch/torch.h>
+#include "c10/core/ScalarType.h"
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/CusparseLtKernels.cpp:6]: Please use angle brackets

File: benchmark_cusparselt.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
+import os
```
Comment (cpuhrsch) [inline benchmark_cusparselt.py:1]: Can you move this into the benchmark folder?

Comment (cpuhrsch) [reply benchmark_cusparselt.py:1]: I think it's generally useful

File: aten/src/ATen/native/cuda/CusparseLtKernels.cpp (hunk 3 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/cuda/CusparseLtKernels.cpp:4]: That means we'd need to move this construct, if it's going to use custom_class, into the torch folder. We can land native functions for the CUDA dispatch key for some of the core routines.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Generally looks good, we just need to make sure to keep aten free of torch dependencies and will thus have to move some code around.

I think the benchmark file is something we could / should put into our benchmark folder (or ao equivalent) and could be generally useful.

File: aten/src/ATen/native/cuda/CusparseLtKernels.cpp (hunk 3 above)
Comment (jcaip) [reply aten/src/ATen/native/cuda/CusparseLtKernels.cpp:4]: Makes sense. Right now we need custom class to keep track of some state like the alg plan and weight matrix descriptor, but I think I might be able to refactor.

File: debug.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
+        if abs(num) < 1024.0:
+            return f"{num:3.1f}{unit}{suffix}"
+        num /= 1024.0
+    return f"{num:.1f}Yi{suffix}"
+
+
+    # run a sweep for the n, batch_size combination
+    # then try running batch size different from the initialized batch size to see effect of caching alg plan.
+    elif args.mode == "alg-id-sweep":
```
Comment (izaitsevfb) [inline debug.py:200]: This doesn't look right.

File: benchmarks/sparse/benchmark_cusparselt.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    dense_measurement = benchmark.Timer(
+        stmt="model(input_tensor)",
+        globals=locals(),
+    ).blocked_autorange()
+
+    dense_output = model(input_tensor)
+
+    # sparsify weights
+    model.linear.weight = nn.Parameter(
+        SemiStructuredSparseTensor(model.linear.weight, contiguous_output=contiguous)
```
Comment (cpuhrsch) [inline benchmarks/sparse/benchmark_cusparselt.py:69]: Can we have a factory function for this? It's more PyTorch-esque (like torch.ones and torch.empty)

File: torch/ao/pruning/semi_structured_sparse_tensor.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+    @staticmethod
+    def __new__(
+        cls,
+        original_tensor,
+        original_shape=None,
+        compressed_tensor=None,
+        cslt=None,
+        transposed=False,
+        contiguous_output=False,
```
Comment (cpuhrsch) [inline torch/ao/pruning/semi_structured_sparse_tensor.py:17]: I don't think contiguous_output should be a user setting. The contiguity is a contract tied to PyTorch operators. If the output of an operation is usually contiguous, then when passing the structured sparse layout it should also be contiguous.

Comment (jcaip) [reply torch/ao/pruning/semi_structured_sparse_tensor.py:17]: Contiguous mode is quite a bit slower than the non-contiguous mode, so I think it's worth having some way to specify which one you want. 

image

This can live somewhere else off of the tensor though.

**Comment by jcaip** [comment]
PR merged as https://github.com/pytorch/pytorch/pull/102135
